asyncapi/v2/v2.go

```diff
@@ -250,12 +257,22 @@ type SubscribeOperation struct {
 	Operation
 }
 
+// NewSubscribeOperation creates a new SubscribeOperation. Useful for testing.
+func NewSubscribeOperation(msg *Message) *SubscribeOperation {
+	return &SubscribeOperation{Operation: *NewOperation(OperationTypeSubscribe, msg)}
+}
+
```

Comment:
Both structs are needed because of the way the decoding works. In order to set the `Type` field on the `Operation` based on the field name (Subscribe field means an Operation of type `Subscribe`, and so on) we need to set the value of that `Type` field as default. That's the purpose of the `MapStructureDefaults` method both structs implement (I created an interface called `MapStructureDefaultsProvider` for this). 

That can't be done in the `Operation` struct since once the decoder is decoding the `Operation` struct, it does not know about the field name it holds that `Operation` struct. 
In summary, whenever an `Operation` struct is being decoded, we "can't" know if it belongs to `Subscribe` or `Publish` field.

Of course we could know, but we would need to build a very customized decoder that does all the particular logic. 
My idea behind not holding the whole raw JSON document on each struct and access the values through that is for pure performance: All this unmarshaling happens on bootstrap time, however accessing the fields is in runtime, so better to have everything in place once the app runs.

Does it make sense?

Reply:
It make sense, so TLD;DR; In Golang simpler is decoding object as:

```bash
`subscribe` -> SubscribeOperation
`publish` -> PublishOperation
```

than

```
`subscribe` -> Operation with `type` field set to `subscribe`
`publish` -> Operation with `type` field set to `publish`
```

Am I right?

Reply:
Yeah, you are right. The point is that I can't tell the decoder to use a default value for the `type` field at this moment. At least the decoders (both json and mapstructure) don't support default values or setting constructors as a way of generating the object.

Reply:
Supporting `discriminators` in golang will be pain 😄  But yeah, I understand now, it can stay.